### ISSUE TITLE
java.lang.InterruptedException  when Thread.sleep(pollTime * 1000L)

### ISSUE THREAD

**Issue by fsfshijm** [jenkinsci/saltstack-plugin#87]
saltstack-plugin version: 3.0.0
The error is:
```
Running jid: 20170504045943435060
Waiting for 1 minions
0 minions are done
Will check status every 5 seconds...
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
java.lang.InterruptedException
	at com.waytta.Builds.runBlockingBuild(Builds.java:180)
	at com.waytta.SaltAPIBuilder.performRequest(SaltAPIBuilder.java:280)
	at com.waytta.SaltAPIStep$SaltAPIStepExecution.run(SaltAPIStep.java:230)
	at com.waytta.SaltAPIStep$SaltAPIStepExecution.run(SaltAPIStep.java:191)
	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousStepExecution.start(AbstractSynchronousStepExecution.java:42)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeStep(DSL.java:184)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeMethod(DSL.java:126)
	at org.jenkinsci.plugins.workflow.cps.CpsScript.invokeMethod(CpsScript.java:108)
	at groovy.lang.GroovyObject$invokeMethod$2.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:151)
	at org.kohsuke.groovy.sandbox.GroovyInterceptor.onMethodCall(GroovyInterceptor.java:21)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:115)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:149)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:146)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:123)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:123)
	at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.methodCall(SandboxInvoker.java:16)
	at WorkflowScript.run(WorkflowScript:10)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:57)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:109)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:82)
	at sun.reflect.GeneratedMethodAccessor308.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.CollectionLiteralBlock$ContinuationImpl.dispatch(CollectionLiteralBlock.java:55)
	at com.cloudbees.groovy.cps.impl.CollectionLiteralBlock$ContinuationImpl.item(CollectionLiteralBlock.java:45)
	at sun.reflect.GeneratedMethodAccessor311.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.PropertyishBlock$ContinuationImpl.get(PropertyishBlock.java:76)
	at com.cloudbees.groovy.cps.LValueBlock$GetAdapter.receive(LValueBlock.java:30)
	at com.cloudbees.groovy.cps.impl.PropertyishBlock$ContinuationImpl.fixName(PropertyishBlock.java:66)
	at sun.reflect.GeneratedMethodAccessor305.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.ConstantBlock.eval(ConstantBlock.java:21)
	at com.cloudbees.groovy.cps.Next.step(Next.java:74)
	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:154)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:18)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:33)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable$1.call(SandboxContinuable.java:30)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.GroovySandbox.runInSandbox(GroovySandbox.java:108)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:30)
	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:165)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:328)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$100(CpsThreadGroup.java:80)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:240)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:228)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:64)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:112)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
Finished: FAILURE
```

The related code:
Builds.java
```
...
        while (numMinionsDone < numMinions) {
            try {
                Thread.sleep(pollTime * 1000L);
            } catch (InterruptedException ex) {
                Thread.currentThread().interrupt();
                // Allow user to cancel job in jenkins interface
                throw new InterruptedException();
            }
...
```

We didn't cancel our Jenkins job,  so what kind of things will make 'sleep' interrupted?

**Comment by fsfshijm** [comment]
By the way, the job always fail when it tooks about 5mins

**Comment by fsfshijm** [comment]
Does someone know this issue?  it's a bit urgent for us. Thanks!

**Comment by mchugh19** [comment]
Hey @fsfshijm, I'm having trouble replicating. 

I created a new freestyle project to do a local call of cmd.run "sleep 305; ls -la" with wait for completion checked. 

Interrupting the sleep call *should* only occur when you actively cancel the build while it is running. Might you have any other plugins installed that timeout long running jobs? Could you post an example of what your job is doing when it is interrupted? Did this only start occuring with the 3.0.0 update? (that release didn't change any of the sleep or interruption logic, so that would be interesting)

```
Started by user anonymous
Building remotely on minion1 in workspace /tmp/workspace/classic salt
Running jid: 20170508055155119839
Waiting for 1 minions
0 minions are done
Will check status every 5 seconds...
Response on cmd.run "sleep 305; ls -la" for master:
[{"master": {"return": "total 20\ndrwx------  3 root root 4096 Mar 25 07:51 .\ndrwxr-xr-x 25 root root 4096 May  8 05:47 ..\n-rw-r--r--  1 root root 3106 Oct 22  2015 .bashrc\n-rw-r--r--  1 root root  148 Aug 17  2015 .profile\ndrwx------  2 root root 4096 Mar 25 07:51 .ssh"}}]
Finished: SUCCESS
```

**Comment by fsfshijm** [comment]
@mchugh19 thanks very much for your response, here is my code

**Comment by mchugh19** [comment]
Ah ha! Seems to be related to pipelines! Investigating...

**Comment by fsfshijm** [comment]
Cool, thanks very much!!!

**Comment by fsfshijm** [comment]
@mchugh19  Seems that interrupted by workflow-cps-plugin, https://github.com/jenkinsci/workflow-cps-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThread.java: 
```
    @Nonnull Outcome runNextChunk() {
        assert program!=null;

        Outcome outcome;

        final CpsThread old = CURRENT.get();
        CURRENT.set(this);

        try (Timeout timeout = Timeout.limit(5, TimeUnit.MINUTES)) {
            LOGGER.log(FINE, "runNextChunk on {0}", resumeValue);
            Outcome o = resumeValue;
            resumeValue = null;
            outcome = program.run0(o);
            if (outcome.getAbnormal() != null) {
                LOGGER.log(FINE, "ran and produced error", outcome.getAbnormal());
            } else {
                LOGGER.log(FINE, "ran and produced {0}", outcome);
            }

            if (outcome.getNormal() instanceof ThreadTask) {
                // if an execution in the thread safepoint is requested, deliver that
                ThreadTask sc = (ThreadTask) outcome.getNormal();
                ThreadTaskResult r = sc.eval(this);
                if (r.resume!=null) {
                    // yield, then keep evaluating the CPS code
                    resumeValue = r.resume;
                } else {
                    // break but with a different value
                    outcome = r.suspend;
                }
            }
        } finally {
            CURRENT.set(old);
        }

```

There is a 5 mins timeout, but I'm not very sure.

Do you have some ideas?

**Comment by mchugh19** [comment]
I am able to replicate this issue with scripted and declarative pipelines.
But there's nothing in the saltapi plugin that should have a 300 sec
timeout. I've asked on the jenkins-dev mailing list to see if there are
suggestions
https://groups.google.com/forum/#!topic/jenkinsci-dev/ojxDProzJAs

**Comment by fsfshijm** [comment]
From the error log above, I think it's due to workflow-cps-plugin:
the error log:
```
java.lang.InterruptedException
	at com.waytta.Builds.runBlockingBuild(Builds.java:180)
	at com.waytta.SaltAPIBuilder.performRequest(SaltAPIBuilder.java:280)
	at com.waytta.SaltAPIStep$SaltAPIStepExecution.run(SaltAPIStep.java:230)
	at com.waytta.SaltAPIStep$SaltAPIStepExecution.run(SaltAPIStep.java:191)
	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousStepExecution.start(AbstractSynchronousStepExecution.java:42)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeStep(DSL.java:184)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeMethod(DSL.java:126)
	at org.jenkinsci.plugins.workflow.cps.CpsScript.invokeMethod(CpsScript.java:108)
...
	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:165)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:328)
```

and the code of org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk, in which, there is a 5 mins timeout:
```
149     /**
150      * Executes CPS code synchronously a little bit more, until it hits
151      * the point the workflow needs to be dehydrated.
152      */
153     @Nonnull Outcome runNextChunk() {
154         assert program!=null;
155
156         Outcome outcome;
157
158         final CpsThread old = CURRENT.get();
159         CURRENT.set(this);
160
161         try (Timeout timeout = Timeout.limit(5, TimeUnit.MINUTES)) {
162             LOGGER.log(FINE, "runNextChunk on {0}", resumeValue);
163             Outcome o = resumeValue;
164             resumeValue = null;
165             outcome = program.run0(o);
166             if (outcome.getAbnormal() != null) {
167                 LOGGER.log(FINE, "ran and produced error", outcome.getAbnormal());
168             } else {
169                 LOGGER.log(FINE, "ran and produced {0}", outcome);
170             }
171
172             if (outcome.getNormal() instanceof ThreadTask) {
173                 // if an execution in the thread safepoint is requested, deliver that
174                 ThreadTask sc = (ThreadTask) outcome.getNormal();
175                 ThreadTaskResult r = sc.eval(this);
176                 if (r.resume!=null) {
177                     // yield, then keep evaluating the CPS code
178                     resumeValue = r.resume;
179                 } else {
180                     // break but with a different value
181                     outcome = r.suspend;
182                 }
183             }
184         } finally {
185             CURRENT.set(old);
186         }
```
In this method, 'program.run0(o)' will invoke salt api step, the duration of which is larger then 5 mins, so will trigger this timeout.

But I don't know why hard code 5 mins here. From this issue: https://issues.jenkins-ci.org/browse/JENKINS-42561,  they don't want to fix it, seems they have some reasons that I don't understand.

So, can you get more informations from the ticket above? and based on this ticket, can we take some actions?

**Comment by fsfshijm** [comment]
@mchugh19   according to https://issues.jenkins-ci.org/browse/JENKINS-42561,  can we think about using 'DurableStep' approach to implement the saltstack plugin?, I'm not sure, just a proposal.

**Comment by mchugh19** [comment]
Thanks for all of your work here, it's been really helpful!

I've updated the query to the jenkins mailing list to get a recommendation. In the jenkins code, I see reference to both durable step or AbstractSynchronousNonBlockingStepExecution, and would like a recommendation.

**Comment by fsfshijm** [comment]
@mchugh19 There is a response from https://groups.google.com/forum/#!topic/jenkinsci-dev/ojxDProzJAs,  what can we do according to Jesse Glick's suggestion?

**Comment by mchugh19** [comment]
Yep! Planning to get to this during the next week. Jesse didn't recommend a complete solution, so this will require investigation and research.

**Comment by fsfshijm** [comment]
Sure, Thanks very very much!!! Looking forward to the next release.

**Comment by fsfshijm** [comment]
@mchugh19 any update for this issue?Thanks!

**Comment by mchugh19** [comment]
Yep! I was traveling a bit recently and started working on this while on planes. This should be handled in #95. I posted the PR to the jenkins-dev mailing list for review, and I also have to add in logic for resuming the job during a jenkins master restart.

**Comment by mchugh19** [comment]
The mentioned branch was worked a bit more and just commited to master. If you are able, please try out [this beta release](https://github.com/jenkinsci/saltstack-plugin/releases/tag/v3.1.0-beta) incorporating these important updates.

**Comment by mingplan** [comment]
@fsfshijm is a colleague of mine. I will use this beta release in our environment to the test function.

**Comment by mingplan** [comment]
@mchugh19  hi, thanks very much for your creative work. The beta release work very well in our environment. The issue about 5 mins timeout don't happen again.It is Ok to release it. Thanks again!

**Comment by fsfshijm** [comment]
@mchugh19 @mingplan Very glad to hear that, Great job!!!

**Comment by mchugh19** [comment]
This was finally corrected in release 3.1. Should be available on the mirrors in the next 24 hours.
